typescript-loader/loader.js

```diff
@@ -0,0 +1,62 @@
+// 💡 It's probably a good idea to put this loader near the end of the chain.
+
+import path from 'node:path';
+
+import tsc from 'typescript';
+
+const tsExts = new Set(['.mts', '.ts', '.tsx']);
+
+export function resolve(specifier, context, nextResolve) {
+  if (!isTypeScriptFile(specifier, context)) { return nextResolve(specifier, context); }
```

Comment:
To support the automatic `.js -> .ts` tsc is doing:
```suggestion
  if (!isTypeScriptFile(specifier, context)) {
    try {
      return nextResolve(specifier, context);
    } catch (err) {
    if (err?.code === "ERR_MODULE_NOT_FOUND") {
     if(url.endsWith(".js")) {
      try {
        return next(url.replace(/\.js$/, ".ts"), context);
      } catch(err2) {
          if (err2) {
            err2.cause = err;
            throw err2;
          }
      }
     } else if (url.endsWith(".mjs")) {
      try {
        return next(url.replace(/\.mjs$/, ".mts"), context);
      } catch(err2) {
          if (err2) {
            err2.cause = err;
            throw err2;
          }
      }
     } else if (url.endsWith(".cjs")) {
      try {
        return next(url.replace(/\.cjs$/, ".cts"), context);
      } catch(err2) {
          if (err2) {
            err2.cause = err;
            throw err2;
          }
      }
     }
    }
    throw err;
  }
```

Reply:
It will blow up in the `load` stage when the raw source still contains `.ts` :(

Reply:
`load` should transpile any file that has a name that ends with `.[m|c]?ts`.

Reply:
ohhh! wait, do you mean to still use `esbuild` (or whatever) to do the transpilation?

```ts
// test.ts
import foo from './foo.js'; // actually foo.mts
```

```js
export function resolve(s, c, n) {
  if (isTS) return {
    format: 'typescript',
    shortCircuit: true,
    url: whatever,
  };
}

export async function load(u, c, n) {
  if (c.format !== 'typescript') return n(…);
  
  // …
  
  const { code: transpiledSource } = esbuild.transform(rawSource, {…});
  
  // …
}
```

Reply:
> ```ts
> // test.ts
> import foo from './foo.js'; // actually foo.mts
> ```

That wouldn't work with `tsc`. It should be

```ts
// test.ts
import foo from './foo.mjs'; // actually foo.mts
```



> do you mean to still use `esbuild` (or whatever) to do the transpilation?

uh I mean why not, but that's not what I suggesting here. I don't think there's any value in having the resolve hook returning `format: 'typescript'`, I would skip that part, and use the file extension instead to detect TS files.

```js
export function resolve(url, context, next) {
  try {
    // First, try to resolve using default resolver:
    return next(url, context);
  } catch (err) {
    // The default resolver failed
    if (err?.code === 'ERR_MODULE_NOT_FOUND') {
      // If the file wasn't found, maybe the user is trying to load a TS file using a JS extension.
      if (url.endsWith('.js')) {
        try {
          // re-try the default resolver, but using `.ts` instead of `.js` as extension
          return next(url.replace(/\.js$/, '.ts'), context);
        } catch (err2) {
          // Nope, that still didn't work, aborting.
          if (err2) {
            err2.cause = err;
            throw err2
          }
        }
      } else if (url.endsWith('.mjs')) { /* same thing but with .mts */ }
      } else if (url.endsWith('.cjs')) { /* same thing but with .cts */ }
    }
    throw err
  }
}

export async function load(url, context, next) {
  if (!/\.[m|c]?ts$/.test(url)) return next(url, context);
  /* transpile TS file somehow */
}
```

Reply:
I think I was able to write this a little more cleanly: d1c27ed

I opted for ``specifier.replace(base, `${name}${tsExt}`)`` instead of your regex because the file extension is not necessarily the final character in the specifier (ex `file:///…/foo.mjs?gen=1`)